### PR TITLE
Script: Add Metadata to ingest context.

### PR DESCRIPTION
**Pull request by stu-elastic**
Adds the `Metadata` class and `metadata()` method to the ingest context.

Metadata has getters and setters for index, id, routing, version and versionType.
It also has a getter for timestamp.

Refs: #86472

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/script/field/MapBackedMetadata.java [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+            if (number.doubleValue() != version) {
+                // did we round?
+                throw new IllegalArgumentException(
+                    "version may only be set to an int or a long but was [" + number + "] with type [" + obj.getClass().getName() + "]"
+                );
+            }
+            return version;
+        }
+        throw new IllegalArgumentException(
+            "version may only be set to an int or a long but was [" + obj + "] with type [" + obj.getClass().getName() + "]"
```
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/script/field/MapBackedMetadata.java:65]: Seeing this exception on a get is confusing. We should perhaps validate the ctx map when it is passed into the ctor?

File: server/src/main/java/org/elasticsearch/script/field/MapBackedMetadata.java [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+    public Object getRawVersion() {
+        return map.get(VERSION);
+    }
+
+    public MapBackedMetadata setVersion(Long version) {
+        return set(VERSION, version);
+    }
+
+    public void removeVersion() {
```
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/script/field/MapBackedMetadata.java:77]: When is it ok to remove version? It's odd having this and getRawVersion, which is different from the rest of the getters/setters.

File: server/src/main/java/org/elasticsearch/script/field/MapBackedMetadata.java [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+        map.remove(VERSION);
+    }
+
+    public String getRouting() {
+        return getString(ROUTING);
+    }
+
+    public MapBackedMetadata setRouting(String routing) {
```
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/script/field/MapBackedMetadata.java:85]: Can we do without chaining? Shouldn't users be doing `=` with these so returning the metadata for chaining seems unnecessary?

**Review by rjernst — COMMENTED** [review]
Looks pretty good, I have some comments.

File: server/src/main/java/org/elasticsearch/script/field/MapBackedMetadata.java (hunk 1 above)
Comment (stu-elastic) [reply server/src/main/java/org/elasticsearch/script/field/MapBackedMetadata.java:65]: The script still has access to ctx and may change it there.  Unfortunately, that means we cannot trust the contents of the map at any point.

File: server/src/main/java/org/elasticsearch/script/field/MapBackedMetadata.java [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+    public Long getVersion() {
+        Object obj = getRawVersion();
+        if (obj == null) {
+            return null;
+        } else if (obj instanceof Number number) {
+            long version = number.longValue();
+            if (number.doubleValue() != version) {
+                // did we round?
+                throw new IllegalArgumentException(
+                    "version may only be set to an int or a long but was [" + number + "] with type [" + obj.getClass().getName() + "]"
```
[the remark this answers is not in the document]
Comment (stu-elastic) [reply server/src/main/java/org/elasticsearch/script/field/MapBackedMetadata.java:59]: The script may put anything it wants in the backing map via ctx.

File: server/src/main/java/org/elasticsearch/script/field/MapBackedMetadata.java (hunk 1 above)
Comment (stu-elastic) [reply server/src/main/java/org/elasticsearch/script/field/MapBackedMetadata.java:65]: Yup, we'd have to override `merge`, `put`, `putAll`, `putIfAbsent`, `replace(k,v)`, `replace(k,v,v)` and `replaceAll`.

Would that be a backwards incompatible change?  It would be rejected after the script runs, but during script execution the map can hold anything as long as it's fixed up before script termination.

Comment (qhoxie) [reply server/src/main/java/org/elasticsearch/script/field/MapBackedMetadata.java:65]: Would the wrapped map block/validate all changes, or just do this same type of validation for the [known metadata fields](https://github.com/elastic/elasticsearch/pull/87309/files/502aac29619e4c4021b69085d584deb1b9a30e7c#diff-6b5913f6e06049c3e3c48eb973883a79554093b08b481006d8af106a92a8680cR18-R22)?

Comment (rjernst) [reply server/src/main/java/org/elasticsearch/script/field/MapBackedMetadata.java:65]: It would block writing to unknown keys. ie the only valid keys are the known metadata fields (the list you linked isn't all of them, it depends on the script context, but those I think are the ones for ingest, or most of them).

Comment (rjernst) [reply server/src/main/java/org/elasticsearch/script/field/MapBackedMetadata.java:65]: Actually, on further inspection of IngestDocument, I think we can do this as I originally suggested @stu-elastic . The sourceAndMetadata map is what we need to wrap. On put, if the key is one of the known metadata fields, we validate the value (this would get validated later anyways when we try to index, this is just moving it more upfront). Unknown keys would just be delegated without validation because those are actually document keys.

Comment (stu-elastic) [reply server/src/main/java/org/elasticsearch/script/field/MapBackedMetadata.java:65]: To help illustrate the BWC issue, the following valid script would become invalid if we validated the map.
```
ctx._version = 'foo';
...
ctx._version = (ctx._version == 'foo') ? 10: 20;
```

File: modules/lang-painless/src/main/resources/org/elasticsearch/painless/org.elasticsearch.script.ingest.meta.txt [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    VersionType getVersionType()
+    void setVersionType(VersionType)
+    ZonedDateTime getTimestamp()
+}
+
+class org.elasticsearch.script.IngestScript {
+    Metadata meta()
+}
+
+class org.elasticsearch.index.VersionType {
```
[the remark this answers is not in the document]
Comment (stu-elastic) [reply modules/lang-painless/src/main/resources/org/elasticsearch/painless/org.elasticsearch.script.ingest.meta.txt:32]: Yes, `VersionType.INTERNAL`.

We could possibly bring them in as constants on the script class, which should make that possible.  It would impact any variable named `internal`, however.

File: modules/lang-painless/src/main/resources/org/elasticsearch/painless/org.elasticsearch.script.ingest.meta.txt [hunk 6]
```diff
@@ -0,0 +1,36 @@
+#
+# Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+# or more contributor license agreements. Licensed under the Elastic License
+# 2.0 and the Server Side Public License, v 1; you may not use this file except
+# in compliance with, at your election, the Elastic License 2.0 or the Server
+# Side Public License, v 1.
+#
+
+# This file only whitelists the script class and the Metadata class for ingest scripts.
+# Due to name conflicts, this whitelist cannot be loaded with other scripts with conflicting Metadata types.
```
[the remark this answers is not in the document]
Comment (stu-elastic) [reply modules/lang-painless/src/main/resources/org/elasticsearch/painless/org.elasticsearch.script.ingest.meta.txt:10]: The painless test context [loads](https://github.com/elastic/elasticsearch/pull/87309/files#diff-ba2678b0d3fb572e0a19426a2dd2c323425afc15b83369a7ff08ca449c803057L111) all script whitelists.  It's not an issue in this change, but when there are multiple Metadata classes, there is a name collision.

The solution of this PR is to avoid loading meta.txt whitelists in painless_test.  Another option is to explicitly name the script whitelists needed by painless test.

File: modules/lang-painless/src/main/resources/org/elasticsearch/painless/org.elasticsearch.script.ingest.meta.txt (hunk 5 above)
[the remark this answers is not in the document]
Comment (stu-elastic) [reply modules/lang-painless/src/main/resources/org/elasticsearch/painless/org.elasticsearch.script.ingest.meta.txt:32]: We could make those available as getters on the script object but that would be a back compat issue with existing variables of the same name.  Scripts would get:

> variable [external] is already defined

We'd have to have a way to shadow to avoid that issue.

File: server/src/main/java/org/elasticsearch/script/field/MapBackedMetadata.java (hunk 1 above)
Comment (tvernum) [reply server/src/main/java/org/elasticsearch/script/field/MapBackedMetadata.java:65]: IIUC, we're only talking about validating the known metadata fields and the user-facing change would be that they're validated every time they're set rather than only being validated at the end of the script (or, when they're read from the context as part of document storage).

That feels OK to me. If we only validate on `getVersion` (etc) then there's a cost to the user that they cannot easily trace where the version was set. They know that something in their script set the version to a string, but that could potentially happen in more than 1 place.

So, yes, there's the possibility that it might break something for someone, but that's true of many bug fixes. This feels like a reasonable feature to validate on set rather than validate on first use. I agree it's borderline, and someone might have a compelling argument for why we should consider it "breaking" rather than "fixing", but I think I'm OK with it.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/PainlessPlugin.java [hunk 7]
```diff
@@ -90,13 +90,13 @@ public final class PainlessPlugin extends Plugin implements ScriptPlugin, Extens
...
             List<Whitelist> contextWhitelists = new ArrayList<>();
-            if (PainlessPlugin.class.getResourceAsStream("org.elasticsearch.script." + context.name.replace('-', '_') + ".txt") != null) {
-                contextWhitelists.add(
-                    WhitelistLoader.loadFromResourceFiles(
-                        PainlessPlugin.class,
-                        "org.elasticsearch.script." + context.name.replace('-', '_') + ".txt"
-                    )
-                );
+            String contextWhitelistName = "org.elasticsearch.script." + context.name.replace('-', '_');
+            if (PainlessPlugin.class.getResourceAsStream(contextWhitelistName + ".txt") != null) {
```
[the remark this answers is not in the document]
Comment (stu-elastic) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/PainlessPlugin.java:94]: https://github.com/elastic/elasticsearch/pull/87539

File: modules/lang-painless/src/main/resources/org/elasticsearch/painless/org.elasticsearch.script.ingest.meta.txt [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+    void setVersion(long)
+    void removeVersion()
+    boolean hasVersion()
+    VersionType getVersionType()
+    void setVersionType(VersionType)
+    ZonedDateTime getTimestamp()
+}
+
+class org.elasticsearch.script.IngestScript {
+    Metadata meta()
```
[the remark this answers is not in the document]
Comment (stu-elastic) [reply modules/lang-painless/src/main/resources/org/elasticsearch/painless/org.elasticsearch.script.ingest.meta.txt]: Done.

File: server/src/main/java/org/elasticsearch/script/IngestScript.java [hunk 9]
```diff
@@ -33,18 +37,100 @@ public abstract class IngestScript {
...
+        this.metadata = metadata;
     }
 
     /** Return the parameters for this script. */
     public Map<String, Object> getParams() {
         return params;
     }
 
-    public abstract void execute(Map<String, Object> ctx);
+    public Map<String, Object> getCtx() {
```
[the remark this answers is not in the document]
Comment (stu-elastic) [reply server/src/main/java/org/elasticsearch/script/IngestScript.java:53]: Added

File: server/src/main/java/org/elasticsearch/script/field/MapBackedMetadata.java [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+
+/** Metadata storage backed by a Map for compatibility with the ctx Map used by ingest and update scripts. */
+public class MapBackedMetadata {
+    public static final String INDEX = "_index";
+    public static final String ID = "_id";
+    public static final String VERSION = "_version";
+    public static final String ROUTING = "_routing";
+    public static final String SOURCE = "_source";
+
+    protected Map<String, Object> map;
```
[the remark this answers is not in the document]
Comment (stu-elastic) [reply server/src/main/java/org/elasticsearch/script/field/MapBackedMetadata.java:33]: Added.

File: server/src/main/java/org/elasticsearch/script/IngestScript.java [hunk 11]
```diff
@@ -33,18 +37,100 @@ public abstract class IngestScript {
...
+        return metadata != null ? metadata.store.getMap() : null;
+    }
+
+    public Metadata meta() {
```
[the remark this answers is not in the document]
Comment (stu-elastic) [reply server/src/main/java/org/elasticsearch/script/IngestScript.java]: Changed.

File: server/src/main/java/org/elasticsearch/script/field/MapBackedMetadata.java [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+
+    public MapBackedMetadata(Map<String, Object> map) {
+        this.map = map;
+    }
+
+    public MapBackedMetadata(int size) {
```
[the remark this answers is not in the document]
Comment (stu-elastic) [reply server/src/main/java/org/elasticsearch/script/field/MapBackedMetadata.java]: Moved to utility.

File: server/src/main/java/org/elasticsearch/script/field/MapBackedMetadata.java (hunk 3 above)
Comment (stu-elastic) [reply server/src/main/java/org/elasticsearch/script/field/MapBackedMetadata.java]: This is used in a future change, removing until then.

**Comment by stu-elastic** [comment]
After discussion with @rjernst, the plan is to
* Remove `org.elasticsearch.script.ingest.meta.txt` and put the metadata stuff in `org.elasticsearch.script.ingest.txt`
* Figure out how to handle `painless_test` loading all whitelists.  This will become an issue for when other metadata classes are whitelisted
* Remove all enums, take strings instead (`VersionType.EXTERNAL` becomes `'external'`)
* `MapBackedMetadata` will implement the `Map<String, Object>` interface and do validation on `put` for `ctx` backwards compatibility.  Only metadata keys will be validated.

File: server/src/main/java/org/elasticsearch/script/field/MapBackedMetadata.java (hunk 2 above)
Comment (stu-elastic) [reply server/src/main/java/org/elasticsearch/script/field/MapBackedMetadata.java:80]: [IngestService](https://github.com/elastic/elasticsearch/blob/a4fdb0ba669cf473a3238faaa28f780d612ff65b/server/src/main/java/org/elasticsearch/ingest/IngestService.java#L908) expects this to be non-null and throws a 500 if it is null.

File: modules/lang-painless/src/main/resources/org/elasticsearch/painless/org.elasticsearch.script.ingest.meta.txt (hunk 5 above)
[the remark this answers is not in the document]
Comment (stu-elastic) [reply modules/lang-painless/src/main/resources/org/elasticsearch/painless/org.elasticsearch.script.ingest.meta.txt]: Changed to taking string.

File: modules/lang-painless/src/main/resources/org/elasticsearch/painless/org.elasticsearch.script.ingest.meta.txt (hunk 6 above)
[the remark this answers is not in the document]
Comment (stu-elastic) [reply modules/lang-painless/src/main/resources/org/elasticsearch/painless/org.elasticsearch.script.ingest.meta.txt]: Removed

**Comment by stu-elastic** [comment]
Added https://github.com/elastic/elasticsearch/pull/87673 for wrapping the `sourceAndMetadata` map.

**Comment by stu-elastic** [comment]
@rjernst I've added https://github.com/elastic/elasticsearch/pull/88102 to enforce non-null `_version`.   This PR has been updated to use the data structure in https://github.com/elastic/elasticsearch/pull/87673.

It's ready for another pass when you have the time.

File: modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/ScriptProcessor.java [hunk 13]
```diff
@@ -120,16 +117,17 @@ public ScriptProcessor create(
...
                 // verify script is able to be compiled before successfully creating processor.
-                IngestScript ingestScript = null;
+                IngestScript.Factory ingestScriptFactory = null;
                 try {
-                    final IngestScript.Factory factory = scriptService.compile(script, IngestScript.CONTEXT);
-                    if (ScriptType.INLINE.equals(script.getType())) {
-                        ingestScript = factory.newInstance(script.getParams());
+                    ingestScriptFactory = scriptService.compile(script, IngestScript.CONTEXT);
+                    if (ScriptType.STORED.equals(script.getType())) {
+                        // do not cache stored scripts lest they change and invalidate the cached value
```
Comment (rjernst) [inline modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/ScriptProcessor.java:124]: Is this a change in current behavior? It might be worth pulling this out into a separate PR, only so that it can be noted (perhaps as a bugfix) in the release notes, whereas if it goes in within this enhancement/addition to the fields API it would be hidden.

File: server/src/main/java/org/elasticsearch/script/IngestScript.java [hunk 14]
```diff
@@ -33,18 +33,31 @@ public abstract class IngestScript {
...
 
     /** Return the parameters for this script. */
     public Map<String, Object> getParams() {
         return params;
     }
 
-    public abstract void execute(Map<String, Object> ctx);
+    /** Provides backwards compatibility access to ctx */
+    public Map<String, Object> getCtx() {
+        return metadata;
```
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/script/IngestScript.java:50]: Can we pass ctx separately to ingest script, so that Metadata can remain strongly typed?

**Review by rjernst — COMMENTED** [review]
This looks much better. A few more thoughts.

File: modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/ScriptProcessor.java (hunk 13 above)
Comment (stu-elastic) [reply modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/ScriptProcessor.java:124]: This is the same as current behavior, it failed a test which is how I found it.

File: server/src/main/java/org/elasticsearch/script/IngestScript.java (hunk 14 above)
Comment (stu-elastic) [reply server/src/main/java/org/elasticsearch/script/IngestScript.java]: Yup, done.

File: modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/ScriptProcessor.java (hunk 13 above)
Comment (stu-elastic) [reply modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/ScriptProcessor.java:124]: This is the current behavior, it's just the instance used to be cached.  This change is due to a failed unit test when I switched to the compiled script rather than the instance.

**Review by rjernst — APPROVED** [review]
